src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1389 @@
+  IL_0112:  ldc.i4.s   -2
+  IL_0114:  stfld      ""int S.<Main>d__0.<>1__state""
+  IL_0119:  ldarg.0
+  IL_011a:  ldflda     ""System.Runtime.CompilerServices.AsyncTaskMethodBuilder S.<Main>d__0.<>t__builder""
+  IL_011f:  call       ""void System.Runtime.CompilerServices.AsyncTaskMethodBuilder.SetResult()""
+  IL_0124:  nop
+  IL_0125:  ret
+}");
+        }
+
+        [Fact]
+        public void TestWithNullableExpression()
+        {
+            string source = @"
+struct S : System.IAsyncDisposable
+{
+    public static async System.Threading.Tasks.Task Main()
+    {
+        S? s = new S();
+        using await (s)
+        {
+            System.Console.Write(""body "");
+            return;
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        System.Console.Write(""DisposeAsync"");
+        return System.Threading.Tasks.Task.CompletedTask;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "body DisposeAsync");
+        }
+
+        [Fact]
+        public void TestWithNullNullableExpression()
+        {
+            string source = @"
+struct S : System.IAsyncDisposable
+{
+    public static async System.Threading.Tasks.Task Main()
+    {
+        S? s = null;
+        using await (s)
+        {
+            System.Console.Write(""body"");
+            return;
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        System.Console.Write(""NOT RELEVANT"");
+        return System.Threading.Tasks.Task.CompletedTask;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            var verifier = CompileAndVerify(comp, expectedOutput: "body");
+        }
+
+        [Fact]
+        public void AssignsInAsyncWithAsyncUsing()
+        {
+            var comp = CreateCompilationWithMscorlib46(@"
+class C
+{
+    public static void M2()
+    {
+        int a=0, x, y, z;
+        L1();
+        a++;
+        x++;
+        y++;
+        z++;
+
+        async void L1()
+        {
+            x = 0;
+            using await (null) { }
+            y = 0;
+
+            // local function exists with a pending branch from the async using, in which `y` was not assigned
+        }
+    }
+}" + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (10,9): error CS0165: Use of unassigned local variable 'y'
+                //         y++;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "y").WithArguments("y").WithLocation(10, 9),
+                // (11,9): error CS0165: Use of unassigned local variable 'z'
+                //         z++;
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "z").WithArguments("z").WithLocation(11, 9)
+                );
```

Comment:
Do we report an error for `await L1(); ...; y++;`?

Reply:
See [https://github.com/dotnet/roslyn/issues/24267](https://github.com/dotnet/roslyn/issues/24267) for that scenario and another related one

---
In reply to: [162762997](https://github.com/dotnet/roslyn/pull/23961#discussion_r162762997) [](ancestors = 162762997)